add cmake exporter for CLion use

thinkberg opened the pull request:
## Description

I am working mainly with CLion and have until now written the CMakelists.txt by hand, which is very uncomfortable. This is an exporter that exports a cmake config which basically adds all the necessary defines and settings, so the IDE knows about the code and adds a custom target that will compile using the mbed-cli. 

## Status
**IN DEVELOPMENT**

I need some input on this. Everything works nicely, but for some reason the Makefile and the mbed-cli links somehow differently than what happens with cmake. Some info on the best way to configure the flags and arguments to the linker would be nice.

run `mbed export -i cmake_gcc_arm` to create the CMakeLists.txt

## Todos

- [ ] add support for other toolchains than GCC_ARM

amq commented:
@thinkberg for debugging and SWO you need to externally use J-Link Debugger and J-Link SWO Viewer, right?

thinkberg commented:
CLion has an in-source debugger view that uses any gdb compatible gdb server. So, yes, JLinkGDBServer will work, as will pyOCD or openocd. You have to start the server in a terminal somewhere and configure a remote debug target, but otherwise, extremly comfortable IMHO. I can post a few screenshots if that is of interest.

amq commented:
@thinkberg I'd be grateful for screenshots

thinkberg commented:
**the toolchain setup (I have added an ARM toolchain)**
<img width="1031" alt="screen shot 2017-11-11 at 21 05 09" src="https://user-images.githubusercontent.com/276461/32693300-0936715c-c729-11e7-8418-cc570bcb7510.png">

**remote debugging setup**
<img width="1066" alt="screen shot 2017-11-11 at 21 11 32" src="https://user-images.githubusercontent.com/276461/32693301-0956c7f4-c729-11e7-823c-a46f18410617.png">

**the target selection (via the cmake exporter)**
It creates a single library target for each `mbed add` library and a special target for the app.
<img width="587" alt="screen shot 2017-11-11 at 21 14 50" src="https://user-images.githubusercontent.com/276461/32693302-0972c260-c729-11e7-9373-382b2fe3f5e1.png">

**I added a `project.cmake` with a target that flashes to the eval board**
<img width="1674" alt="screen shot 2017-11-11 at 21 21 10" src="https://user-images.githubusercontent.com/276461/32693303-0992b282-c729-11e7-9ec5-d6eb290b38e0.png">

**the view after running remote debug and halting at a breakpoint**
the lower right window shows the JLinkGDBServer output: `JLinkGDBServer -device nRF52 -if swd -port 3333`
<img width="1679" alt="screen shot 2017-11-11 at 21 22 40" src="https://user-images.githubusercontent.com/276461/32693304-09b89d9e-c729-11e7-8bda-6030c7cf411a.png">

**stopping at a more interesting point**
You can select variables and see their value or inspect them in the source editor. Also it displays values of variables directly in the source at line 212.
<img width="1677" alt="screen shot 2017-11-11 at 21 39 01" src="https://user-images.githubusercontent.com/276461/32693305-09d71e72-c729-11e7-938b-e64910f165e6.png">

I haven't yet contacted Jetbrains for things like running the gdb server automatically etc. I guess its possible.

theotherjimmy commented on tools/export/cmake/__init__.py:
```diff
@@ -0,0 +1,225 @@
+    def build(project_name, log_name="build_log.txt", cleanup=True):
+        """ Build Make project """
+        # > Make -j
+        cmd = ["make", "-j"]
```
Is this the correct command to run? I was under the impression that you would need to call `cmake` to use a `CMakeLists.txt`.

thinkberg replied:
There are parts of the code that are not obvious to me, maybe you can help me understand:

When is this build called, except for cleanup?

And, actually, you do run cmake only once and then make or ninja etc. I will need to find a way to identify what is needed (like `cmake -E make`, but this requires that I know where the cmake build dir is in the end. I will try to find something that works.

theotherjimmy replied:
The Mbed OS handbook contains [documentation of the `bulid` method](https://github.com/ARMmbed/Handbook/blob/new_engine/docs/reference/contributing/tools/exporter.md#the-build-static-method). It's pretty sparse, so feel free to ask questions, and I'll answer them and update that document.

thinkberg replied:
[WIP]

theotherjimmy commented on tools/export/cmake/__init__.py:
```diff
@@ -0,0 +1,225 @@
+
+    @classmethod
+    def is_target_supported(cls, target_name):
+        return True
```
I know this is incorrect. You don't support every target unconditionally. Could you update this to check for support much like the makefile exporers do?

thinkberg replied:
fixed.

theotherjimmy commented on tools/export/cmake/__init__.py:
```diff
@@ -0,0 +1,225 @@
+        sources = set(self.resources.c_sources + \
+                      self.resources.cpp_sources + \
+                      self.resources.s_sources + \
+                      self.resources.headers)
```
I don't think that headers are sources. Maybe this variable should be renamed?

thinkberg replied:
I did explain in the code, adding headers helps IDEs detect files belonging to the project. cmake ignores them. As a purist I don't like it, but it makes life easier.

theotherjimmy commented on tools/export/cmake/__init__.py:
```diff
@@ -0,0 +1,225 @@
+        libnames = [l[:-4] for l in self.resources.lib_refs]
+        libs = {re.sub(r'^[.]/', '', l): sorted([f for f in sources if f.startswith(l)]) for l in libnames}
+        libs = {k: v for k, v in libs.items() if len(v) != 0}
+        srcs = sorted([f for f in sources if f not in [item for sublist in libs.values() for item in sublist]])
```
This looks like it might be a check for uniqueness. `set()`, used above, already guarantees uniqueness, so I don't think this is needed. Further, it's very hard to follow exactly what this is doing.

thinkberg replied:
I will take a look and update the PR. Some of your comments obviously relay to my lazyness...

thinkberg replied:
explanation in the code. python has not subtract list from list, so its done like this.

theotherjimmy commented on tools/export/cmake/__init__.py:
```diff
@@ -0,0 +1,225 @@
+                      self.resources.s_sources + \
+                      self.resources.headers)
+
+        libnames = [l[:-4] for l in self.resources.lib_refs]
```
despite the fact that these are `.lib` files, you might want to name these variable something other than `lib*` as they can easily be confused with `.a` handling. Something like dependencies or deps might be better.

thinkberg replied:
fixed.

theotherjimmy commented on tools/export/cmake/__init__.py:
```diff
@@ -0,0 +1,225 @@
+        else:
+            ctx['pp'] = None
+
+        for templatefile in ['cmake/%s.tmpl' % self.TEMPLATE]:
```
Why are you looping through a list size one? Could you simplify this?

thinkberg replied:
I will take a look and update the PR. Some of your comments obviously relay to my lazyness...

thinkberg replied:
fixed

theotherjimmy commented on tools/export/cmake/__init__.py:
```diff
@@ -0,0 +1,242 @@
+        # we flatten the list of source files from all dependencies and
+        # then only add file to srcs if its not in that list
+        # (basically srcs = allSourcefiles - flatten(depSources.values())
+        srcs = [f for f in allSourceFiles if f not in [item for sublist in depSources.values() for item in sublist]]
```
I think this may be easier with a set. For example:
```python
srcs = set(allSourceFiles) - sum(set(dep) for dep in depSources.values(), set())
```

I think you mentioned something about python lists not supported subtraction. Python sets do. Maybe this was the logic you were looking for.

thinkberg replied:
I found an even simpler and more readable way with your hint.

thinkberg commented:
More progress, my project now builds.

- For some reason the hex files after objcopy (without softdevice, nrf52) differe between mbed compile and cmake build
- the linking appears to be fragile, after sorting the static dependencies in a way libmbed-os.a is last it worked, before there were issues not finding mbed functions
- currently the exporter uses the profile at export time. cmake would support different profiles if there was a way to get all the different command line options for each of the profiles

I still need to work on the build() for the CI.

0xc0170 commented:
@thinkberg Please rebase properly, commits from master should not be part of this patch (visible here)

thinkberg commented:
Sorry. Also for the delay, had to fix some retina detachment first...

theotherjimmy commented:
@thinkberg No worries. I hope you're felling better!

cmonr commented:
Is this PR still in progress?

thinkberg commented:
Yes, still on it, just other work that got inbetween...

thinkberg commented:
I am having linker issues. All compiles fine, but the last step fails:

```
/usr/local/Cellar/gcc-arm-none-eabi/20160928/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/lib/armv7e-m/softfp/libc.a(lib_a-signalr.o): In function `_kill_r':
signalr.c:(.text._kill_r+0x10): undefined reference to `_kill'
/usr/local/Cellar/gcc-arm-none-eabi/20160928/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/lib/armv7e-m/softfp/libc.a(lib_a-signalr.o): In function `_getpid_r':
signalr.c:(.text._getpid_r+0x0): undefined reference to `_getpid'
```

Sounds similar to this issue: https://github.com/gnu-mcu-eclipse/eclipse-plugins/issues/58 However, I wonder how to fix it.

thinkberg commented:
Got it working, by adding dependencies for `stdc++`, `c`, and `nosys` (well, actually all system deps) to the final target. The leads to cmake adding them *after* the targets.

Feel free to test and review.

*One issue left is: some of the libraries built will have a few too many source files if they have sub-dependencies. However, this does not seem to influence the built.*

thinkberg commented:
still need to work on the [WIP] above. Almost forgot about it.

theotherjimmy commented:
@thinkberg Thanks for getting to this! It looks like it's coming along quite well.

thinkberg commented:
After finding the command line to run the examples test:

`python ../tools/test/examples/examples.py export cmake_gcc_arm`

I added the required steps to build. It works for me with the blinky example.

thinkberg commented:
I get errors like this when doing the test builds:

`/usr/local/Cellar/gcc-arm-none-eabi/20160928/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: cannot find -lublox-odin-w2-driver`

Is this expected?

theotherjimmy commented:
> I get errors like this when doing the test builds:
```
/usr/local/Cellar/gcc-arm-none-eabi/20160928/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: cannot find -lublox-odin-w2-driver
```
>Is this expected?


No build failures are expect.

It looks like you may need to add `resources.lib_dirs` knowlage to your exporter. For reference [the makefile template](https://github.com/ARMmbed/mbed-os/blob/master/tools/export/makefile/Makefile.tmpl#L56) add these resources [to the jinja context](https://github.com/ARMmbed/mbed-os/blob/master/tools/export/makefile/__init__.py#L78).

thinkberg commented:
added the resources libs feature, the library is now found

cmonr commented:
@theotherjimmy Do you think you could do a re-review of the changes?

/morph build

thinkberg commented:
Found a small issue: separating the libs into their own cmake targets breaks linking sometimes, due to the fact that the individual targets do not explicitely depend on each other, like a lib that depends also on mbed-os (which often is the case), but is appended to the link-libraries after the mbed-os lib. Cmake can't make assumptions there and the dependencies are not ordered as such.

@theotherjimmy Is there a way to see the dependencies?

One way to fix this would be to add mbed-os as to be linked to all dependent libs. Alternatively I could make the whole cmake project flat, which somehow is against my Sheldon-esque sense of order.

off-topic: I am going to be on London Wed/Thu, but I guess nobody is close for a coffee/beer?

theotherjimmy commented:
@thinkberg There is not a way to see dependencies. We do not have that information anywhere. You may want to have a special case to put mbed-os last. 

-------

It sounds like you are building your exporter in a different way than mbed OS is built:
Your exporter: every library is built as a library and linked; Mbed OS: every C/C++ file is built independently and linked. Switching to the Mbed OS way of building would certainly solve the issue.

thinkberg commented:
okay, maybe flatten the whole thing and be done with it then.

thinkberg commented:
Flattened, and made sure the include paths are depth first, so automatic includes by clion won't create strange long include statements.

thinkberg commented:
@theotherjimmy I hope I got it now.

theotherjimmy commented:
@thinkberg Yep. You got it!

theotherjimmy commented:
I'm going to lunch. I'll review this PR when I get back.

0xc0170 commented:
> I'm going to lunch. I'll review this PR when I get back.

And has not returned yet 🍴⏲😆

thinkberg commented:
Lunch coma. Must have been awesome food.

theotherjimmy commented on tools/export/cmake/CMakeLists.txt.tmpl:
```diff
@@ -0,0 +1,106 @@
+# optional custom target to build via mbed-cli
+ADD_CUSTOM_TARGET(mbed-cli-build
+        COMMAND ${CMAKE_COMMAND} -E echo "mbed compile --build BUILD/${CMAKE_BUILD_TYPE} --profile ${MBED_BUILD_PROFILE}"
+        COMMAND mbed compile --build BUILD/${CMAKE_BUILD_TYPE} --profile ${MBED_BUILD_PROFILE}
```
Is this the primary compiling method?

thinkberg replied:
No, that’s just for convenience. The primary method is to choose the {{name}} target.

theotherjimmy replied:
Thanks.

cmonr commented:
/morph build

cmonr commented:
/morph build

cmonr commented:
@thinkberg, is it fair to say that this PR will not be getting export functionality for other toolchains?

thinkberg commented:
Not this one. However, it should be easy to do one has access to armcc etc. I just didn’t have the time.

cmonr commented:
That's fine. Just wanted to make sure.

cmonr commented:
/morph build

0xc0170 commented:
/morph build

drewcassidy commented:
I know this has ben closed and merged but I'm having a few issues with it
- it generates a getting started html file every time you update the cmake file
- the cmake-build-debug directory seems to confuse the compiler. I was having errors where it couldnt find `BLE.h` because it was searching in the cmake build directory instead. Maybe this should automatically add `cmake-build-debug` to `.mbedignore`?

cmonr commented:
@drewcassidy I would suggest opening an issue, and linking it to this PR. We generally stop paying attention to pull request comments after they've been merged.

thinkberg commented:
Please open an issue and I will take a look. The GettingStarted.html is generated by the exporter every time. Not sure how to stop that. And a hint as how to add a line safely to .mbedignore would be nice.

theotherjimmy commented:
> And a hint as how to add a line safely to .mbedignore would be nice.

You could also generate a `.mbedignore` file containing:
```
*
```
within `cmake-build-debug`, avoiding the whole safety question.

elmot commented:
@amq, @thinkberg, FIY, here is a plugin for CLion + OpenOCD
https://plugins.jetbrains.com/plugin/10115-openocd--stm32cubemx-support-for-arm-embedded-development
